Add user option to specify which initial residual norm to compare against

jwpeterson opened the pull request:
This PR is an initial attempt to give people more options when working with `PresetNodalBCs`, hopefully allowing them to avoid using the dreaded `nl_abs_tol`. The PR adds a new parameter in the Executioner section called `use_initial_residual_before_preset_bcs` (the name is not final) which defaults to true --  this is the existing behavior of MOOSE.  If you set it to false, MOOSE will instead use the residual norm which is computed _after_ the PresetNodalBC values have been applied to the solution vector in order to determine convergence.  This option should only affect examples which use PresetBCs.

The new option is demonstrated in `tensor_mechanics/tests/mohr_coulomb/uni_axial1.i`.  In this example, the old `nl_abs_tol=1.e-4` is now replaced with `nl_rel_tol=1.e-10`.  Since @WilkAndy also independently fixed a bug in this test in #5030, I'm not too worried about the small CSV diffs, I'll iron them out now that his PR has been merged into devel.

jwpeterson committed:
Adding option to use residual before *or* after setting PresetBC values.

Refs #5030.

jwpeterson committed:
Update the Mohr-Coulomb uni_axial1.i test to use nl_rel_tol and the new option.

Since we changed the convergence test from absolute to relative, the
solution changed slightly, so I also regolded.  @WilkAndy may want to
take a look and see if the result still seems reasonable.

moosebuild commented:
Results of testing 6159d540c6a967980debb4ab124cd81e6a023b6f using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15928

moosebuild commented:
Results of testing 6159d540c6a967980debb4ab124cd81e6a023b6f using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15931

moosebuild commented:
Results of testing 6159d540c6a967980debb4ab124cd81e6a023b6f using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15929

moosebuild commented:
Results of testing 6159d540c6a967980debb4ab124cd81e6a023b6f using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/15930

jwpeterson committed:
Rename use_initial_residual_before_preset_bcs -> compute_initial_residual_before_preset_bcs

jwpeterson committed:
Print residual norm in NonlinearSystem::solve() before PresetBCs have been applied.

This restores an old MOOSE feature, and is necessary for avoiding
confusion when relative convergence is being detected based on this
residual rather than the residual printed by PETSc.

Refs #5030.

moosebuild commented:
Results of testing a9a625de18bf9d2310f02c6942ffea4b59cd27ba using moose_PR_pre_check recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16010

moosebuild commented:
Results of testing a9a625de18bf9d2310f02c6942ffea4b59cd27ba using moose_PR_app_tests recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16013

moosebuild commented:
Results of testing a9a625de18bf9d2310f02c6942ffea4b59cd27ba using moose_PR_test recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16011

moosebuild commented:
Results of testing a9a625de18bf9d2310f02c6942ffea4b59cd27ba using moose_PR_test_dbg recipe:

Passed on: linux-gnu

View the results here: https://www.moosebuild.com/view_job/16012

permcody commented:
@bwspenc, @cpritam -You might want to take a look at this too. This new option will allow you to "fiddle" with which initial residual used in the convergence checks when using PresetBCs. Ideally the default for MOOSE will be to use always use the residuals from the Presets in calculating the initial residual. Once we get all of the solid mechanics tests using this new option, we'll switch the default in MOOSE.

If you have feedback on this, let us know.
